### DISCUSSION TITLE
DataGrid and TreeDataGrid Copy & Paste

### DISCUSSION THREAD

**Discussion by timunie** [AvaloniaUI/Avalonia#8591] [Ideas]
## Current Situation:

Currently the copy & paste options in DataGrid is quite limited. Atm you can only copy in CSV Format with `\t` as separator and `\r\n` as new line. If it comes to advanced needs, you are limited. Examples:
- You have a program which uses `,` or `;` as separator ► You will need to convert it manually
- You have a multi line text column with line breaks ► Your output is broken
- You want to include `" ... "` ► Your output is broken
- ...

Moreover maybe one would need the data in a different format like: 
- HTML
- RTF
- JSON
- MarkDown

## Idea 💡:

My idea is to add an interface and / or a base class called `ICopyDataTable` (we can discuss it's name) which is a `StyledProperty` in DataGrid and TreeDataGrid. In there we have some methods and Properties which must be implemented: 
```cs
- string GetClipboardContent(IEnumerable<IEnumerable<object>> dataTable); // returns the final string for the Clipboard
- string PrepareCell(object? cellContent); // preperares the cell content (for example escape special chars)
- string PrepareRow(IEnumerable<object> rowContent); // preperares the row content (for example escape special chars)
- string PrepareHeaderRow(IEnumerable<object> rowContent); // preperares the header row content (for example escape special chars, formatting)
- string? HeaderText {get; set;} // a header text like <html>-open tag
- string? FooterText {get; set;} // a footer text like </html>-end tag
- string? CellPrefix {get; set;} // any cell opening text like <td>-open tag
- string? CellSuffix {get; set;} // any cell closing text like </td>-end tag
- string? RowPrefix {get; set;} // any row opening text like <tr>-open tag
- string? RowSuffix {get; set;} // any row closing text like </tr>-end tag
- string? CellsDelimiter {get; set;} // a delimiter to add between cells like \t or ;
- string? RowsDelimiter {get; set;} // a delimiter to add between rows like \n\r
```

Let me know your thoughts

**Comment by jp2masa** [comment]
Honestly I don't think this belongs to the UI framework. In my opinion, this is part of the model layer for most UI architectures. For the really simple case where you have a data grid and don't have a model layer, and just want to export the data, I could see it being useful, but how many people will actually use that?

**Reply by timunie** [reply]
CSV is already implemented. The other formats can be discussed.

**Reply by wosiyourfather** [reply]
Do you have an example

**Reply by timunie** [reply]
Not public as it is a hack-fix kind of ;-)

**Comment by Flithor** [comment]
Wait... No copy method can use in code behind. It's only provide for Ctrl+C?
And.. Wait... 
![image](https://github.com/user-attachments/assets/09466948-3e82-4ac5-9b3c-a7184cdb96d1)
`Insert` key for ... copy?

**Reply by rabbitism** [reply]
That's Ctrl+insert

**Reply by Flithor** [reply]
> That's Ctrl+insert

As I know.. Ctrl+Insert means: Paste.
But Google tell me.. it's... copy?

-----------

Back to the point.
In the end, I write the copy code by myself.
The copy function doesn't provide a public method, only the shortcut entry.

I don't know why it was designed this way.. or is it just... forgotten?
